### PR TITLE
Sprint2

### PR DESCRIPTION
**Pull request by twollnik**

### CODE REVIEW

File: src/lib/storage/chunk.hpp [hunk 1]
```diff
@@ -25,9 +25,15 @@ class Chunk : private Noncopyable {
...
   Chunk() = default;
 
+  // Creates a chunk with the segments vector initialized to the specified capacity
+  explicit Chunk(ColumnID n_columns);
+
   // Adds a segment to the "right" of the chunk.
   void add_segment(const std::shared_ptr<AbstractSegment> segment);
 
+  // Exchanges a segment at the specified index
+  void add_segment_at(const std::shared_ptr<AbstractSegment> segment, const ColumnID position);
```
Comment (LukasBudach) [inline src/lib/storage/chunk.hpp:35]: Naming could be more descriptive/accurate if the function was called ``insert_segment_at(...)`` as this does not necessarily add a new segment, but may also insert a segment, replacing an already existing one.

File: src/lib/storage/dictionary_segment.cpp [hunk 2]
```diff
@@ -1,24 +1,62 @@
...
 template <typename T>
 DictionarySegment<T>::DictionarySegment(const std::shared_ptr<AbstractSegment>& abstract_segment) {
-  // Implementation goes here
+
+  // determine unique values and store in sorted set
+  auto dict_values = std::set<T>{};
+  auto segment_size = abstract_segment->size();
+  for (auto value_index = ChunkOffset{0}; value_index < segment_size; ++value_index) {
+    dict_values.emplace(type_cast<T>((*abstract_segment)[value_index]));
+  }
```
Comment (LukasBudach) [inline src/lib/storage/dictionary_segment.cpp:21]: You could do the sorting + making unique of the segment values without introducing an additional data structure in the same time complexity by pushing all values into ``_dictionary``and then using ``std::unique``, ``_dictionary.erase()`` and ``_dictionary.shrink_to_fit()``.

File: src/lib/storage/dictionary_segment.cpp [hunk 3]
```diff
@@ -1,24 +1,62 @@
...
+  _dictionary.reserve(n_unique_values);
+  auto cur_dict_index = ValueID{0};
+  for (const auto& value : dict_values) {
+    _dictionary.push_back(value);
+    dict_indexes[value] = cur_dict_index;
+    cur_dict_index++;
+  }
+
+  // apply dictionary encoding to input segment
+  for (ChunkOffset value_index = 0; value_index < segment_size; ++value_index) {
```
Comment (LukasBudach) [inline src/lib/storage/dictionary_segment.cpp:46]: Use auto-to-stick:
```suggestion
  for (auto value_index = ChunkOffset{0}; value_index < segment_size; ++value_index) {
```

File: src/lib/storage/dictionary_segment.cpp [hunk 4]
```diff
@@ -1,24 +1,62 @@
...
+    auto cur_value = type_cast<T>((*abstract_segment)[value_index]);
+    _attribute_vector->set(value_index, dict_indexes[cur_value]);
+  }
 }
 
 template <typename T>
 AllTypeVariant DictionarySegment<T>::operator[](const ChunkOffset chunk_offset) const {
-  // Implementation goes here
-  return AllTypeVariant{};
+  return AllTypeVariant{_dictionary[_attribute_vector->get(chunk_offset)]};
```
Comment (LukasBudach) [inline src/lib/storage/dictionary_segment.cpp:54]: You should not need the cast to ``AllTypeVariant`` here.

File: src/lib/storage/dictionary_segment.cpp [hunk 5]
```diff
@@ -1,24 +1,62 @@
...
 }
 
 template <typename T>
 T DictionarySegment<T>::get(const ChunkOffset chunk_offset) const {
-  // Implementation goes here
-  return T{};
+  return _dictionary[_attribute_vector->get(chunk_offset)];
```
Comment (LukasBudach) [inline src/lib/storage/dictionary_segment.cpp:59]: Since ``get`` should probably do a range check (in contrast to ``operator[]``, we recomment using the ``value_of_value_id`` function here.

File: src/lib/storage/dictionary_segment.cpp [hunk 6]
```diff
@@ -28,61 +66,62 @@ void DictionarySegment<T>::append(const AllTypeVariant& value) {
...
+  return _dictionary;
 }
 
 template <typename T>
-std::shared_ptr<const AbstractAttributeVector> DictionarySegment<T>::attribute_vector() const {
-  // Implementation goes here
-  return nullptr;
+std::shared_ptr<AbstractAttributeVector> DictionarySegment<T>::attribute_vector() const {
+  return _attribute_vector;
 }
```
Comment (LukasBudach) [inline src/lib/storage/dictionary_segment.cpp:75]: Why did you remove the ``const`` in the return type (was: ``std::shared_ptr<const AbstractAttributeVector>``)?

File: src/lib/storage/dictionary_segment.cpp [hunk 7]
```diff
@@ -28,61 +66,62 @@ void DictionarySegment<T>::append(const AllTypeVariant& value) {
...
-  // Implementation goes here
-  return ValueID{};
+  return upper_bound(type_cast<T>(value));
 }
 
 template <typename T>
 ChunkOffset DictionarySegment<T>::unique_values_count() const {
-  // Implementation goes here
-  return ChunkOffset{};
+  return static_cast<ChunkOffset>(_dictionary.size());
```
Comment (LukasBudach) [inline src/lib/storage/dictionary_segment.cpp:112]: There is no need for the explicit cast here, as there is no compiler warning for a narrowing cast.

File: src/lib/storage/dictionary_segment.cpp [hunk 8]
```diff
@@ -28,61 +66,62 @@ void DictionarySegment<T>::append(const AllTypeVariant& value) {
...
 }
 
 template <typename T>
 ChunkOffset DictionarySegment<T>::size() const {
-  // Implementation goes here
-  return ChunkOffset{};
+  return static_cast<ChunkOffset>(_attribute_vector->size());
```
Comment (LukasBudach) [inline src/lib/storage/dictionary_segment.cpp:117]: There is no need for the explicit cast here, as there is no compiler warning for a narrowing cast.

File: src/lib/storage/fixed_width_integer_vector.cpp [hunk 9]
```diff
@@ -0,0 +1,37 @@
+#include "fixed_width_integer_vector.hpp"
+
+#include "utils/assert.hpp"
+#include "types.hpp"
+
+namespace opossum {
+
+template <typename uintX_t>
+FixedWidthIntegerVector<uintX_t>::FixedWidthIntegerVector(const size_t capacity) 
+  : _vector{std::vector<uintX_t>(capacity)} {};
```
Comment (LukasBudach) [inline src/lib/storage/fixed_width_integer_vector.cpp:10]: The parameter being called ``capacity`` is not great, as it actually contains the number of elements expected to be pushed into the vector and not the capacity, i.e., memory space, required.

File: src/lib/storage/fixed_width_integer_vector.cpp [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+template <typename uintX_t>
+size_t FixedWidthIntegerVector<uintX_t>::size() const {
+  return _vector.size();
+}
+
+template <typename uintX_t>
+AttributeVectorWidth FixedWidthIntegerVector<uintX_t>::width() const {
+  return static_cast<AttributeVectorWidth>(sizeof(uintX_t));
```
Comment (LukasBudach) [inline src/lib/storage/fixed_width_integer_vector.cpp:30]: We would remove the ``static_cast`` here, as the compiler does not seem to mind the cast from ``size_t`` to ``uint8_t`` at this point.

File: src/lib/storage/fixed_width_integer_vector.cpp [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+template class FixedWidthIntegerVector<uint8_t>;
+template class FixedWidthIntegerVector<uint16_t>;
+template class FixedWidthIntegerVector<uint32_t>;
```
Comment (LukasBudach) [inline src/lib/storage/fixed_width_integer_vector.cpp:35]: An alternative would be:
`BOOST_PP_SEQ_FOR_EACH(EXPLICIT_INSTANTIATION, FixedWidthIntegerVector, (uint8_t)(uint16_t)(uint32_t))`

File: src/lib/storage/fixed_width_integer_vector.hpp [hunk 12]
```diff
@@ -0,0 +1,23 @@
+#pragma once
+
+#include "abstract_attribute_vector.hpp"
+#include "types.hpp"
+
+namespace opossum {
+
+// TODO: comments
```
Comment (LukasBudach) [inline src/lib/storage/fixed_width_integer_vector.hpp:8]: Yes, still is TODO 😄

File: src/lib/storage/table.cpp [hunk 13]
```diff
@@ -75,8 +76,35 @@ std::shared_ptr<Chunk> Table::get_chunk(ChunkID chunk_id) { return _chunks.at(ch
 std::shared_ptr<const Chunk> Table::get_chunk(ChunkID chunk_id) const { return _chunks.at(chunk_id); }
 
 void Table::compress_chunk(const ChunkID chunk_id) {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+
+  DebugAssert(chunk_id < chunk_count(), "invalid chunk id "+std::to_string(chunk_id)+". table only has "+std::to_string(chunk_count())+" chunks");
+
+  auto old_chunk = get_chunk(chunk_id);
```
Comment (LukasBudach) [inline src/lib/storage/table.cpp:82]: ``get_chunk()`` will already perform a range check, so you don't need the assert here (except for the error message being a little nicer I guess)

File: src/lib/storage/table.cpp [hunk 14]
```diff
@@ -75,8 +76,35 @@ std::shared_ptr<Chunk> Table::get_chunk(ChunkID chunk_id) { return _chunks.at(ch
...
+  };
+
+  auto threads = std::vector<std::thread>();
+  for (auto column_index = ColumnID{0}; column_index < n_segments; ++column_index) {
+    auto worker = std::thread(compression_worker_lambda, column_index);
+    threads.emplace_back(std::move(worker));
+  }
+  for (size_t thread_id = 0; thread_id < n_segments; ++thread_id) {
+    threads[thread_id].join();
+  }
```
Comment (LukasBudach) [inline src/lib/storage/table.cpp:103]: This could loop through the vector elements instead of its index
`for (auto&& thread : threads)`...

File: src/lib/storage/table.cpp [hunk 15]
```diff
@@ -75,8 +76,35 @@ std::shared_ptr<Chunk> Table::get_chunk(ChunkID chunk_id) { return _chunks.at(ch
...
+
+  // swap in new dict encoded chunk
+  // TODO(all): consider concurrent accesses when exchanging the chunk?
+  _chunks[chunk_id] = new_chunk;
```
Comment (LukasBudach) [inline src/lib/storage/table.cpp:107]: We have used mutexes throughout the *table.cpp* file to ensure thread-safety in case of concurrent access to chunks/data in the table.

``std::vector`` objects are thread-safe for multiple concurrent reads or a single write, but not for a single write plus any other access.
Therefore, whenever we want to touch the ``_chunks`` vector, we now need to create a lock on it, as there is the possibility of it being written to while we access it.

File: src/test/storage/dictionary_segment_test.cpp [hunk 16]
```diff
@@ -70,6 +70,127 @@ TEST_F(StorageDictionarySegmentTest, LowerUpperBound) {
   EXPECT_EQ(dict_segment->upper_bound(15), INVALID_VALUE_ID);
 }
 
-// TODO(student): You should add some more tests here (full coverage would be appreciated) and possibly in other files.
+TEST_F(StorageDictionarySegmentTest, AccessingElements) {
```
Comment (LukasBudach) [inline src/test/storage/dictionary_segment_test.cpp:73]: We believe it may not be optimal to test multiple functions (``operator[]`` and ``get``) in the same test, as you lose visibility of where the error actually originated.

File: src/test/storage/dictionary_segment_test.cpp [hunk 17]
```diff
@@ -70,6 +70,127 @@ TEST_F(StorageDictionarySegmentTest, LowerUpperBound) {
...
+
+TEST_F(StorageDictionarySegmentTest, UInt8) {
+  for (int32_t index = 0; index < 100; ++index) {
+    value_segment_int->append(index);
+  }
+  value_segment_int->append(42);
+  auto column = std::make_shared<DictionarySegment<int32_t>>(value_segment_int);
+  auto dict_col = std::dynamic_pointer_cast<opossum::DictionarySegment<int32_t>>(column);
+
+  EXPECT_EQ(dict_col->estimate_memory_usage(), 100 * sizeof(int32_t) + 101 * sizeof(uint8_t));
```
Comment (LukasBudach) [inline src/test/storage/dictionary_segment_test.cpp:165]: It would be better to explicitly test the ``estimate_memory_usage`` function, just to ensure that it does not contain errors which then taint the test results here.

File: src/test/storage/dictionary_segment_test.cpp [hunk 18]
```diff
@@ -70,6 +70,127 @@ TEST_F(StorageDictionarySegmentTest, LowerUpperBound) {
...
+TEST_F(StorageDictionarySegmentTest, UInt32) {
+  for (int32_t index = 0; index < 65536; ++index) {
+    value_segment_int->append(index);
+  }
+
+  auto column = std::make_shared<DictionarySegment<int32_t>>(value_segment_int);
+  auto dict_col = std::dynamic_pointer_cast<opossum::DictionarySegment<int32_t>>(column);
+
+  EXPECT_EQ(dict_col->estimate_memory_usage(), 65536 * sizeof(int32_t) + 65536 * sizeof(int32_t));
+  EXPECT_THROW(dict_col->append(0), std::exception);
```
Comment (LukasBudach) [inline src/test/storage/dictionary_segment_test.cpp:193]: You already tested that the ``append()`` function should not work on *DictionarySegment* objects; you can remove that here, as it doesn't really fit into the test anyways.

File: src/test/storage/table_test.cpp [hunk 19]
```diff
@@ -77,6 +78,36 @@ TEST_F(StorageTableTest, GetColumnIdByName) {
...
+
+  table.append({4, "Hello"});
+  table.append({6, "Hello"});
+  table.append({3, "!"});
+
+  table.compress_chunk(ChunkID{0});
+
+  auto encoded_chunk = table.get_chunk(ChunkID{0});
+  auto dict_seg_0 = static_cast<DictionarySegment<int>>(encoded_chunk->get_segment(ColumnID{0}));
+  auto dict_seg_1 = static_cast<DictionarySegment<std::string>>(encoded_chunk->get_segment(ColumnID{1}));
```
Comment (LukasBudach) [inline src/test/storage/table_test.cpp:91]: At this point, we recommend testing whether the compressed chunk actually contains segments of type *DictionarySegment* before static casting them and simply assuming that this works.

We found out, that a ``dynamic_cast`` will throw an ``std::bad_cast`` error if you try to cast a *ValueSegment* into a *DictionarySegment*, so using this we were able to check whether the segment type changed with the ``EXPECT_THROW`` and ``EXPECT_NO_THROW`` checks.

**Review by LukasBudach — CHANGES_REQUESTED** [review]

**Comment by LukasBudach** [comment]
You're missing a test/tests for the ``FixedWidthIntegerVector``.

File: src/lib/storage/dictionary_segment.cpp (hunk 4 above)
Comment (janlemcke) [reply src/lib/storage/dictionary_segment.cpp:54]: Although we do not need the cast, we stick to the explicit casting, to be consistent.

File: src/lib/storage/dictionary_segment.cpp (hunk 7 above)
Comment (janlemcke) [reply src/lib/storage/dictionary_segment.cpp:112]: We stick to this cast because of consistency

File: src/lib/storage/table.cpp (hunk 13 above)
Comment (janlemcke) [reply src/lib/storage/table.cpp:82]: We go with the nicer error message and the debugAssert because it will not be executed in the final release build.

File: src/lib/storage/fixed_width_integer_vector.cpp (hunk 11 above)
Comment (tjeyy) [reply src/lib/storage/fixed_width_integer_vector.cpp:35]: Great alternative, but we'd like to stick with our implementation because it doesn't need any thrid-party libs.

File: src/lib/storage/table.cpp (hunk 15 above)
Comment (tjeyy) [reply src/lib/storage/table.cpp:107]: Good point! We were arguing about the behavior of a vector when exposed to concurrent access.
This behavior makes locking necessary of course.

File: src/test/storage/table_test.cpp (hunk 19 above)
Comment (tjeyy) [reply src/test/storage/table_test.cpp:91]: Makes sense.

**Comment by tjeyy** [comment]
Thank you all for your feedback!
Since we adapted most of your suggestions in sprint 3, we're merging this PR now.

File: src/lib/storage/dictionary_segment.cpp [hunk 20]
```diff
@@ -1,24 +1,62 @@
...
+
+  // determine unique values and store in sorted set
+  auto dict_values = std::set<T>{};
+  auto segment_size = abstract_segment->size();
+  for (auto value_index = ChunkOffset{0}; value_index < segment_size; ++value_index) {
+    dict_values.emplace(type_cast<T>((*abstract_segment)[value_index]));
+  }
+
+  // select the right attribute vector integer type
+  auto n_unique_values = dict_values.size();
```
Comment (d-justen) [inline src/lib/storage/dictionary_segment.cpp:24]: Could be const

File: src/lib/storage/dictionary_segment.cpp [hunk 21]
```diff
@@ -1,24 +1,62 @@
...
+  } else if (n_unique_values <= std::numeric_limits<uint16_t>::max()) {
+    _attribute_vector = std::make_shared<FixedWidthIntegerVector<uint16_t>>(segment_size);
+  } else {
+    _attribute_vector = std::make_shared<FixedWidthIntegerVector<uint32_t>>(segment_size);
+  }
+
+  // build dictionary and store dictionary indexes in hash map for quick lookup during encoding
+  auto dict_indexes = std::unordered_map<T, ValueID>{};
+  dict_indexes.reserve(n_unique_values);
+  _dictionary.reserve(n_unique_values);
```
Comment (d-justen) [inline src/lib/storage/dictionary_segment.cpp:37]: Abbreviations are not consistent. (`dict` vs `dictionary`)

File: src/lib/storage/fixed_width_integer_vector.hpp (hunk 12 above)
Comment (d-justen) [inline src/lib/storage/fixed_width_integer_vector.hpp:8]: dangling todo

File: src/lib/CMakeLists.txt [hunk 22]
```diff
@@ -1,6 +1,8 @@
 # Sources and libraries shared among the different builds of the lib
 set(
     SOURCES
+    storage/fixed_width_integer_vector.hpp
+    storage/fixed_width_integer_vector.cpp
```
Comment (d-justen) [inline src/lib/CMakeLists.txt:5]: order alphabetically

File: src/lib/storage/chunk.hpp (hunk 1 above)
Comment (d-justen) [inline src/lib/storage/chunk.hpp:35]: Method name can be confusing as it sounds as if you are inserting a segment (instead of swapping)

File: src/lib/storage/dictionary_segment.cpp [hunk 23]
```diff
@@ -1,24 +1,62 @@
...
+  auto dict_values = std::set<T>{};
+  auto segment_size = abstract_segment->size();
+  for (auto value_index = ChunkOffset{0}; value_index < segment_size; ++value_index) {
+    dict_values.emplace(type_cast<T>((*abstract_segment)[value_index]));
+  }
+
+  // select the right attribute vector integer type
+  auto n_unique_values = dict_values.size();
+  Assert(n_unique_values <= std::numeric_limits<uint32_t>::max(), "Too many unique values");
+  if (n_unique_values <= std::numeric_limits<uint8_t>::max()) {
```
Comment (d-justen) [inline src/lib/storage/dictionary_segment.cpp:26]: <= does not leave space for INVALID_VALUE_ID (in this case == std::numeric_limits<uint8_t>::max())

File: src/lib/storage/chunk.cpp [hunk 24]
```diff
@@ -17,8 +17,15 @@
 
 namespace opossum {
 
+Chunk::Chunk(ColumnID n_columns) : _segments(std::vector<std::shared_ptr<AbstractSegment>>(n_columns)) {};
```
Comment (dey4ss) [inline src/lib/storage/chunk.cpp:20]: Naming of parameter, avoid abbreviations. Furthermore, initializing your `_segments` vector with `n` elements (aka `nullptr`s) leads to non-intuitive behavior when adding segments with `add_segment()`.

File: src/lib/storage/dictionary_segment.cpp [hunk 25]
```diff
@@ -1,24 +1,62 @@
-#include "dictionary_segment.hpp"
+#include <set>
+#include <unordered_map>
+#include <limits>
```
Comment (dey4ss) [inline src/lib/storage/dictionary_segment.cpp:3]: lexicographical order

File: src/lib/storage/dictionary_segment.cpp [hunk 26]
```diff
@@ -1,24 +1,62 @@
...
+#include "type_cast.hpp"
 
 namespace opossum {
 
 template <typename T>
 DictionarySegment<T>::DictionarySegment(const std::shared_ptr<AbstractSegment>& abstract_segment) {
-  // Implementation goes here
+
+  // determine unique values and store in sorted set
+  auto dict_values = std::set<T>{};
```
Comment (dey4ss) [inline src/lib/storage/dictionary_segment.cpp:17]: `dictionary_values`, `unique_values`, or just `values`

File: src/lib/storage/dictionary_segment.cpp [hunk 27]
```diff
@@ -1,24 +1,62 @@
...
+  auto segment_size = abstract_segment->size();
+  for (auto value_index = ChunkOffset{0}; value_index < segment_size; ++value_index) {
+    dict_values.emplace(type_cast<T>((*abstract_segment)[value_index]));
```
Comment (dey4ss) [inline src/lib/storage/dictionary_segment.cpp:20]: You could resolve your input segment / data type first and omit casting each value. If you had resolved your input segment to a  pointer to a `ValueSegment<T>`, you could do the following:
```c++
  const auto& input_values = value_segment->values();
  const auto unique_values = std::set<T>{input_values.cbegin(), input_values.cend()};
```

File: src/lib/storage/dictionary_segment.cpp (hunk 20 above)
Comment (dey4ss) [inline src/lib/storage/dictionary_segment.cpp:24]: naming

File: src/lib/storage/dictionary_segment.cpp [hunk 28]
```diff
@@ -1,24 +1,62 @@
...
+    _attribute_vector = std::make_shared<FixedWidthIntegerVector<uint16_t>>(segment_size);
+  } else {
+    _attribute_vector = std::make_shared<FixedWidthIntegerVector<uint32_t>>(segment_size);
+  }
+
+  // build dictionary and store dictionary indexes in hash map for quick lookup during encoding
+  auto dict_indexes = std::unordered_map<T, ValueID>{};
+  dict_indexes.reserve(n_unique_values);
+  _dictionary.reserve(n_unique_values);
+  auto cur_dict_index = ValueID{0};
```
Comment (dey4ss) [inline src/lib/storage/dictionary_segment.cpp:38]: naming

File: src/lib/storage/dictionary_segment.cpp [hunk 29]
```diff
@@ -1,24 +1,62 @@
...
+  for (const auto& value : dict_values) {
+    _dictionary.push_back(value);
+    dict_indexes[value] = cur_dict_index;
+    cur_dict_index++;
+  }
+
+  // apply dictionary encoding to input segment
+  for (ChunkOffset value_index = 0; value_index < segment_size; ++value_index) {
+    auto cur_value = type_cast<T>((*abstract_segment)[value_index]);
```
Comment (dey4ss) [inline src/lib/storage/dictionary_segment.cpp:47]: naming. As above, you could omit redundant casts of each value when casting the segment itself.

File: src/lib/storage/dictionary_segment.cpp (hunk 5 above)
Comment (dey4ss) [inline src/lib/storage/dictionary_segment.cpp:59]: Check that `chunk_offset` is not out of bounds.

File: src/lib/storage/dictionary_segment.cpp [hunk 30]
```diff
@@ -28,61 +66,62 @@ void DictionarySegment<T>::append(const AllTypeVariant& value) {
...
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return _dictionary;
 }
 
 template <typename T>
-std::shared_ptr<const AbstractAttributeVector> DictionarySegment<T>::attribute_vector() const {
-  // Implementation goes here
-  return nullptr;
+std::shared_ptr<AbstractAttributeVector> DictionarySegment<T>::attribute_vector() const {
```
Comment (dey4ss) [inline src/lib/storage/dictionary_segment.cpp:73]: Removing the `const` qualifier allows the caller of this accessor to manipulate your attribute vector.

File: src/lib/storage/dictionary_segment.cpp [hunk 31]
```diff
@@ -28,61 +66,62 @@ void DictionarySegment<T>::append(const AllTypeVariant& value) {
...
+  return _attribute_vector;
 }
 
 template <typename T>
 const T DictionarySegment<T>::value_of_value_id(const ValueID value_id) const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
```
Comment (dey4ss) [inline src/lib/storage/dictionary_segment.cpp:44]: You can add an `Assert` to ensure the `value_id` is contained in the dictionary. Doing so, you get an explicit error (with file/line number) instead of a vector range error.

File: src/lib/storage/dictionary_segment.cpp [hunk 32]
```diff
@@ -28,61 +66,62 @@ void DictionarySegment<T>::append(const AllTypeVariant& value) {
...
+  return _dictionary.at(value_id);
 }
 
 template <typename T>
 ValueID DictionarySegment<T>::lower_bound(const T value) const {
-  // Implementation goes here
-  return ValueID{};
+  auto iter = std::lower_bound(_dictionary.begin(), _dictionary.end(), value);
```
Comment (dey4ss) [inline src/lib/storage/dictionary_segment.cpp:84]: `const`?

File: src/lib/storage/dictionary_segment.cpp [hunk 33]
```diff
@@ -28,61 +66,62 @@ void DictionarySegment<T>::append(const AllTypeVariant& value) {
...
+  if (iter == _dictionary.end()) {
+    return INVALID_VALUE_ID;
+  }
+  return static_cast<ValueID>(std::distance(_dictionary.begin(), iter));
```
Comment (dey4ss) [inline src/lib/storage/dictionary_segment.cpp:88]: `.cend()`, `.cbegin()`

File: src/lib/storage/dictionary_segment.cpp [hunk 34]
```diff
@@ -28,61 +66,62 @@ void DictionarySegment<T>::append(const AllTypeVariant& value) {
...
 
 template <typename T>
 ValueID DictionarySegment<T>::lower_bound(const AllTypeVariant& value) const {
-  // Implementation goes here
-  return ValueID{};
+  return lower_bound(type_cast<T>(value));
 }
 
 template <typename T>
 ValueID DictionarySegment<T>::upper_bound(const T value) const {
```
Comment (dey4ss) [inline src/lib/storage/dictionary_segment.cpp:97]: see above

File: src/lib/storage/dictionary_segment.cpp [hunk 35]
```diff
@@ -28,61 +66,62 @@ void DictionarySegment<T>::append(const AllTypeVariant& value) {
...
-  return ChunkOffset{};
+  return static_cast<ChunkOffset>(_attribute_vector->size());
 }
 
 template <typename T>
 size_t DictionarySegment<T>::estimate_memory_usage() const {
-  return size_t{};
+  auto dict_size = sizeof(T) * dictionary().size();
+  auto att_vec_size = attribute_vector()->width() * attribute_vector()->size();
+  return dict_size + att_vec_size;
```
Comment (dey4ss) [inline src/lib/storage/dictionary_segment.cpp:124]: naming, `const`

File: src/lib/storage/dictionary_segment.hpp [hunk 36]
```diff
@@ -8,12 +8,11 @@
 
 #include "abstract_segment.hpp"
 #include "all_type_variant.hpp"
+#include "abstract_attribute_vector.hpp"
```
Comment (dey4ss) [inline src/lib/storage/dictionary_segment.hpp:11]: lexicographical order

File: src/lib/storage/fixed_width_integer_vector.cpp [hunk 37]
```diff
@@ -0,0 +1,37 @@
...
+
+namespace opossum {
+
+template <typename uintX_t>
+FixedWidthIntegerVector<uintX_t>::FixedWidthIntegerVector(const size_t capacity) 
+  : _vector{std::vector<uintX_t>(capacity)} {};
+
+template <typename uintX_t>
+ValueID FixedWidthIntegerVector<uintX_t>::get(const size_t index) const {
+  return static_cast<ValueID>(_vector.at(index));
```
Comment (dey4ss) [inline src/lib/storage/fixed_width_integer_vector.cpp:14]: Range check for meaningful error message (`Assert`)

File: src/lib/storage/table.cpp [hunk 38]
```diff
@@ -11,6 +11,7 @@
 #include <vector>
 
 #include "value_segment.hpp"
+#include "dictionary_segment.hpp"
```
Comment (dey4ss) [inline src/lib/storage/table.cpp:14]: order

File: src/lib/storage/table.cpp [hunk 39]
```diff
@@ -75,8 +76,35 @@ std::shared_ptr<Chunk> Table::get_chunk(ChunkID chunk_id) { return _chunks.at(ch
 std::shared_ptr<const Chunk> Table::get_chunk(ChunkID chunk_id) const { return _chunks.at(chunk_id); }
 
 void Table::compress_chunk(const ChunkID chunk_id) {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+
+  DebugAssert(chunk_id < chunk_count(), "invalid chunk id "+std::to_string(chunk_id)+". table only has "+std::to_string(chunk_count())+" chunks");
```
Comment (dey4ss) [inline src/lib/storage/table.cpp:80]: format (spaces around `+`)

File: src/lib/storage/table.cpp [hunk 40]
```diff
@@ -75,8 +76,35 @@ std::shared_ptr<Chunk> Table::get_chunk(ChunkID chunk_id) { return _chunks.at(ch
...
+
+  auto old_chunk = get_chunk(chunk_id);
+  auto n_segments = old_chunk->size();
```
Comment (dey4ss) [inline src/lib/storage/table.cpp:83]: naming

File: src/lib/storage/table.cpp [hunk 41]
```diff
@@ -75,8 +76,35 @@ std::shared_ptr<Chunk> Table::get_chunk(ChunkID chunk_id) { return _chunks.at(ch
...
+  auto new_chunk = std::make_shared<Chunk>(ColumnID{n_segments});
+
+  auto compression_worker_lambda = [this, &old_chunk, &new_chunk](const ColumnID column_id) {
```
Comment (dey4ss) [inline src/lib/storage/table.cpp:86]: naming

File: src/lib/storage/table.cpp [hunk 42]
```diff
@@ -75,8 +76,35 @@ std::shared_ptr<Chunk> Table::get_chunk(ChunkID chunk_id) { return _chunks.at(ch
...
+    const auto& segment = old_chunk->get_segment(column_id);
+    const auto& type = this->column_type(column_id);
+    resolve_data_type(type, [&](const auto data_type_t) {
+      using ColumnDataType = typename decltype(data_type_t)::type;
+      const auto dictionary_segment = std::make_shared<DictionarySegment<ColumnDataType>>(segment);
+      new_chunk->add_segment_at(dictionary_segment, column_id);
+    });
+  };
+
+  auto threads = std::vector<std::thread>();
```
Comment (dey4ss) [inline src/lib/storage/table.cpp:96]: curly braces?

File: src/lib/storage/table.cpp [hunk 43]
```diff
@@ -75,8 +76,35 @@ std::shared_ptr<Chunk> Table::get_chunk(ChunkID chunk_id) { return _chunks.at(ch
...
+  for (auto column_index = ColumnID{0}; column_index < n_segments; ++column_index) {
+    auto worker = std::thread(compression_worker_lambda, column_index);
+    threads.emplace_back(std::move(worker));
```
Comment (dey4ss) [inline src/lib/storage/table.cpp:99]: `.emplace_back()` constructs the element in-place, so you can simplify this to
```c++
threads.emplace_back(compression_worker_lambda, column_index);
```

File: src/test/storage/dictionary_segment_test.cpp [hunk 44]
```diff
@@ -70,6 +70,127 @@ TEST_F(StorageDictionarySegmentTest, LowerUpperBound) {
...
+  auto dict_segment = std::dynamic_pointer_cast<DictionarySegment<int32_t>>(segment);
+
+  // test [] operator
+  EXPECT_EQ((*dict_segment)[ChunkOffset{0}], AllTypeVariant{1});
+  EXPECT_EQ((*dict_segment)[ChunkOffset{1}], AllTypeVariant{2});
+
+  // test get function
+  EXPECT_EQ(dict_segment->get(ChunkOffset{0}), 1);
+  EXPECT_EQ(dict_segment->get(ChunkOffset{1}), 2);
+  EXPECT_ANY_THROW(dict_segment->get(ChunkOffset{100}));
```
Comment (dey4ss) [inline src/test/storage/dictionary_segment_test.cpp:91]: You could use `EXPECT_THROW` and the specific error class.

File: src/test/storage/dictionary_segment_test.cpp [hunk 45]
```diff
@@ -70,6 +70,127 @@ TEST_F(StorageDictionarySegmentTest, LowerUpperBound) {
...
+TEST_F(StorageDictionarySegmentTest, AccessingElements) {
+  value_segment_int->append(1);
+  value_segment_int->append(2);
+
+  std::shared_ptr<AbstractSegment> segment;
+  resolve_data_type("int", [&](auto type) {
+    using Type = typename decltype(type)::type;
+    segment = std::make_shared<DictionarySegment<Type>>(value_segment_int);
+  });
+  auto dict_segment = std::dynamic_pointer_cast<DictionarySegment<int32_t>>(segment);
```
Comment (dey4ss) [inline src/test/storage/dictionary_segment_test.cpp:82]: You could provide a constructor for a dictionary segment that takes the attribute vector and the dictionary as arguments to omit encoding in the tests. Encoding could be tested separately.

File: src/test/storage/dictionary_segment_test.cpp [hunk 46]
```diff
@@ -70,6 +70,127 @@ TEST_F(StorageDictionarySegmentTest, LowerUpperBound) {
...
+
+  EXPECT_ANY_THROW(dict_segment->append(AllTypeVariant{5}));
```
Comment (dey4ss) [inline src/test/storage/dictionary_segment_test.cpp:105]: specific error class, you can also omit inserting initial values
